### PR TITLE
Improve MPT fp8

### PR DESCRIPTION
**Pull request by atakaha**
Add Softmax and FusedSDPA
Fix unnecessary args  from self._gradient_checkpointing_func() call.

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by atakaha** [comment]
- observing new token difference between orinal and use_flash_attention. Because of SDPA implementation difference, specially attn_mask, between MPT and itorch.nn.functional.scaled_dot_product_attention,

- observing new token difference between original and softmax. This could be cause by precision difference, since softmax call torch.ops.hpu.softmax_fp8

- bf16 sample command line (use softmax) 
```bash
python run_generation.py --model_name_or_path mosaicml/mpt-7b --use_hpu_graphs --use_kv_cache --limit_hpu_graphs --max_input_tokens 128 --max_new_tokens 128 --batch_size 256 --bf16 
```
- bf16 with flash attention sample command line (doesn't use softmax) 
```bash
python run_generation.py --model_name_or_path mosaicml/mpt-7b --use_hpu_graphs --use_kv_cache --limit_hpu_graphs --max_input_tokens 128 --max_new_tokens 128 --batch_size 256 --bf16 --use_flash_attention
```

- throughput sample  

|config | batch size \|max input tokens \| max new tokens | Throughput (including tokenization)(tokens/s) | HPU graphs | Memory allocated(GB) | Max memory allocated(GB) | TP improvement vs bf16|
:--: | :--: | --: | --: | --: | --: | --: |
org bf16 | 128 \| 128 \| 128 | 5076 | 14 | 45.3 | 83.51 | N/A
bf16 + flash_attention | ditto | 5150 | 14 | 46.3 | 84.48 | 1.4%
bf16 + softmax | ditto | 5148 | 14 | 46.3 | 84.48 | 1.4%
fp8 + flash_attention | ditto | 7091 | 21 | 40.1 | 78.29 | 39.7%
fp8 + softmax | ditto | 6921 | 22 | 38.85 | 77.04 | 36.3%
org bf16 | 32 \| 128 \| 1024 | 1604 | 14 | 33.31 | 54.79 | N/A
bf16 + flash_attention | ditto | 1700 | 14 | 35.84 | 57.32 | 1.5%
bf16 + softmax | ditto | 1726 | 14 | 33.31 | 54.79 | 1.5%
fp8 + flash_attention | ditto | 2038 | 21 | 29.64 | 51.12 | 19.9%
fp8 + softmax | ditto | 2084 | 22 | 27.12 | 48.62 | 22.6%

File: optimum/habana/transformers/models/mpt/modeling_mpt.py [hunk 1]
```diff
@@ -300,6 +360,8 @@ def prepare_inputs_for_generation(
         - support for internal bucketing
         """
         bucket_internal = kwargs.get("bucket_internal")
+        use_flash_attention = kwargs.get("use_flash_attention")
+        flash_attention_recompute = kwargs.get("flash_attention_recompute")
```
Comment (tthakkal) [inline optimum/habana/transformers/models/mpt/modeling_mpt.py:364]: minor
```suggestion
        use_flash_attention = kwargs.get("use_flash_attention", False)
        flash_attention_recompute = kwargs.get("flash_attention_recompute", False)
```

File: optimum/habana/transformers/models/mpt/modeling_mpt.py [hunk 2]
```diff
@@ -21,142 +21,196 @@
...
+        hidden_states: torch.Tensor,
+        position_bias: torch.Tensor,
+        past_key_value: Optional[Tuple[torch.Tensor]] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        token_idx: Optional[torch.Tensor] = None,
+        use_flash_attention: Optional[bool] = False,
+        flash_attention_recompute: Optional[bool] = False,
+    ):
+        """
+        Copied from MptAttention.forward: https://github.com/huggingface/transformers/blob/v4.32.0/src/transformers/models/mpt/modeling_mpt.py
```
[the remark this answers is not in the document]
Comment (atakaha) [reply optimum/habana/transformers/models/mpt/modeling_mpt.py:70]: This line is original, line 123

File: optimum/habana/transformers/models/mpt/modeling_mpt.py [hunk 3]
```diff
@@ -244,6 +300,8 @@ def forward(
                     use_cache,
                     output_attentions,
                     None,
+                    use_flash_attention,
```
[the remark this answers is not in the document]
Comment (atakaha) [reply optimum/habana/transformers/models/mpt/modeling_mpt.py:303]: removed newly added 2 params.

[the remark this answers is not in the document]
Comment (tthakkal) [reply optimum/habana/transformers/models/mpt/modeling_mpt.py:303]: I see the code exists still

[the remark this answers is not in the document]
Comment (atakaha) [reply optimum/habana/transformers/models/mpt/modeling_mpt.py]: removed these and pushed.

File: optimum/habana/transformers/models/mpt/modeling_mpt.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tthakkal) [reply optimum/habana/transformers/models/mpt/modeling_mpt.py:70]: still please update to latest as we are copying code from latest

[the remark this answers is not in the document]
Comment (atakaha) [reply optimum/habana/transformers/models/mpt/modeling_mpt.py]: merge MptAttention forward r4.44.1

**Review by tthakkal — APPROVED** [review]

**Comment by atakaha** [comment]
@regisss, @mandy-li , Please review this PR.

**Comment by mandy-li** [comment]
@atakaha , pls use make style to fix code format issue

**Comment by atakaha** [comment]
> @atakaha , pls use make style to fix code format issue

make style issue is fixed.

**Review by tthakkal — APPROVED** [review]

**Comment by mandy-li** [comment]
@libinta , please have somebody review this PR.thanks

File: optimum/habana/transformers/models/mpt/modeling_mpt.py [hunk 4]
```diff
@@ -15,180 +15,207 @@
...
+        else:
+            attention_scores = torch.matmul(query_states, key_states.transpose(-1, -2)) * self.softmax_scale
+
+            if position_bias is not None:
+                attention_scores = attention_scores + position_bias
+            if attention_mask is not None:
+                attention_scores = attention_scores.masked_fill(attention_mask, torch.finfo(query_states.dtype).min)
+
+            # (batch_size, n_heads, seq_length, key_length)
+            if self.is_fp8:
```
Comment (jiminha) [inline optimum/habana/transformers/models/mpt/modeling_mpt.py:146]: I think this part should be taken care of internally to use fp8 or bf16 kernel based on the setting. Did you actually see the fp8 kernel is called during bf16 run without this code? 
 
Falcon example. 
https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/models/falcon/modeling_falcon.py#L180

File: optimum/habana/transformers/models/mpt/modeling_mpt.py [hunk 5]
```diff
@@ -15,180 +15,207 @@
...
+                    value_states,
+                    attention_mask * torch.finfo(query_states.dtype).min + position_bias.to(query_states.dtype),
+                    0.0,
+                    False,
+                    None,
+                )
+            attn_weights = None
+        else:
+            attention_scores = torch.matmul(query_states, key_states.transpose(-1, -2)) * self.softmax_scale
+
```
Comment (jiminha) [inline optimum/habana/transformers/models/mpt/modeling_mpt.py:139]: You should also use fp8 matmul kernel. Please check falcon code below. 
https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/models/falcon/modeling_falcon.py#L108

File: optimum/habana/transformers/models/mpt/modeling_mpt.py [hunk 6]
```diff
@@ -15,180 +15,207 @@
...
+            position_bias_query_index = max(0, position_bias.size(1) - query_length)
+            position_bias_key_index = max(0, position_bias.size(2) - key_length)
+
+            position_bias = position_bias[:, position_bias_query_index:, position_bias_key_index:]
+
+        if use_flash_attention and FusedSDPA:
+            import habana_frameworks.torch.hpu as ht
+
+            with ht.sdp_kernel(enable_recompute=flash_attention_recompute):
```
Comment (jiminha) [inline optimum/habana/transformers/models/mpt/modeling_mpt.py:126]: Can you also check other model if this enable_recompute should be set based on the fp8/bf16 and q_len.

File: optimum/habana/transformers/models/mpt/modeling_mpt.py (hunk 5 above)
Comment (atakaha) [reply optimum/habana/transformers/models/mpt/modeling_mpt.py:139]: I tested it and throughput slowdown. The trace showed if MPT use it then it doesn't use fp8, but torch.matmul use fp8. That's the why MPT doesn't use the fp8 matmul kernel this time.

File: optimum/habana/transformers/models/mpt/modeling_mpt.py (hunk 6 above)
Comment (atakaha) [reply optimum/habana/transformers/models/mpt/modeling_mpt.py:126]: If set enable_recompute same way of other models, then FP8 throughput drop to half. In the trace, softmax_stage1_fwd_f32 appears and spending a lot of time, There is not this process at enable_recompute =  False case.

File: optimum/habana/transformers/models/mpt/modeling_mpt.py (hunk 4 above)
Comment (atakaha) [reply optimum/habana/transformers/models/mpt/modeling_mpt.py]: I confirmed softmax_fp8 take care fp8 and bf16 cases. code is updated.

File: optimum/habana/transformers/models/mpt/modeling_mpt.py (hunk 5 above)
Comment (jiminha) [reply optimum/habana/transformers/models/mpt/modeling_mpt.py:139]: That's weird.  Wondering if fp8_sofmax actually forcing matmul to also run with fp8.  Do you see a good accuracy with fp8 softmax?

File: optimum/habana/transformers/models/mpt/modeling_mpt.py (hunk 6 above)
Comment (jiminha) [reply optimum/habana/transformers/models/mpt/modeling_mpt.py:125]: what's your fp8 command with flash_attention?

File: optimum/habana/transformers/models/mpt/modeling_mpt.py (hunk 5 above)
Comment (atakaha) [reply optimum/habana/transformers/models/mpt/modeling_mpt.py:139]:
```
class Matmul(torch.nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, x, y):
        return torch.matmul(x, y)
```

Compare profile data this between FP8, for GPT-J, Mistral and MPT.

- GPT-J and Mistral spend <70% for MME but MPT spend ~30%.
- GPT-J and Mistral calling index_copy_fwd_hf8 and cast_bf16_to_hf8 kernels but doesn't appear these kernel call in MPT. When add matmul_qk and matmul_qv in blocklist of maxabs_quant.json for MPT then it call same kernels of GPT-J/Mistral.
This is reason we didn't add this.

File: optimum/habana/transformers/models/mpt/modeling_mpt.py (hunk 6 above)
Comment (jiminha) [reply optimum/habana/transformers/models/mpt/modeling_mpt.py:125]: Did you test any case which enable causal_mask and enable_compute? 
How about longer prompt? Usually causal_mask shows better perf for long prompt

File: optimum/habana/transformers/models/mpt/modeling_mpt.py (hunk 5 above)
Comment (jiminha) [reply optimum/habana/transformers/models/mpt/modeling_mpt.py:139]: Also, can you make sure your change doesn't break anything for the training case since model file is used for both.

File: optimum/habana/transformers/models/mpt/modeling_mpt.py (hunk 6 above)
Comment (atakaha) [reply optimum/habana/transformers/models/mpt/modeling_mpt.py:125]: > what's your fp8 command with flash_attention?
The command line is 
```
QUANT_CONFIG=./quantization_config/maxabs_quant.json \
python run_generation.py \
--model_name_or_path mosaicml/mpt-7b \
--use_hpu_graphs \
--use_kv_cache \
--limit_hpu_graphs \
--max_input_tokens 128 \
--max_new_tokens 128 \
--batch_size 128 \
--bf16 \
--use_flash_attention
```

Comment (atakaha) [reply optimum/habana/transformers/models/mpt/modeling_mpt.py:125]: > Did you test any case which enable causal_mask and enable_compute? How about longer prompt? Usually causal_mask shows better perf for long prompt

No, I haven't test it these cases.

File: optimum/habana/transformers/models/mpt/modeling_mpt.py [hunk 7]
```diff
@@ -15,180 +15,204 @@
...
-        attention_scores = torch.matmul(query_states, key_states.transpose(-1, -2)) * self.softmax_scale
+logger = logging.get_logger(__name__)
+
+
+class Softmax(nn.Module):
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, x, dim=None, invAttnHead=None):
+        return torch.ops.hpu.softmax_fp8(x, dim, None, None, invAttnHead)
```
Comment (tthakkal) [inline optimum/habana/transformers/models/mpt/modeling_mpt.py:52]: Since INC is enabled, please use `torch.nn.functional.softmax` as its supported module by INC for quantization.
https://docs.habana.ai/en/latest/PyTorch/Inference_on_PyTorch/Inference_Using_FP8.html#supported-functions
```suggestion
        return torch.nn.functional.softmax(x, dim)
```

Comment (atakaha) [reply optimum/habana/transformers/models/mpt/modeling_mpt.py]: updated the code with your suggestion.

**Review by jiminha — APPROVED** [review]
Looks good. Could you investigate further though to see if we need causal_mask and recompute enabled differently for long prompt?

**Review by tthakkal — APPROVED** [review]

**Comment by atakaha** [comment]
> The code quality check failed, please run `make style`.

fixed ruff error.

**Comment by atakaha** [comment]
> Looks good. Could you investigate further though to see if we need causal_mask and recompute enabled differently for long prompt and submit as a separate patch if needed?

Sure, I will.

**Comment by atakaha** [comment]
@regisss, Please review this PR.

**Review by regisss — APPROVED** [review]
